plugin/folding_range.py

```diff
@@ -0,0 +1,138 @@
+
+    def _handle_response_async(self, change_count: int, response: Optional[List[FoldingRange]]) -> None:
+        self.change_count = change_count
+        self.folding_ranges = response or []
+
+    def description(
+        self, manual: bool = True, hidden: bool = False, event: Optional[dict] = None, point: Optional[int] = None
+    ) -> str:
+        if manual:
+            return "LSP: Fold"
+        # Implementation detail of Sublime Text: TextCommand.description is called *before* TextCommand.is_visible
+        self.folding_region = None
+        if self.change_count != self.view.change_count():  # Ensure that the response has already arrived
+            return "LSP <debug>"  # is_visible will return False
+        if point is not None:
+            pt = point
+        else:
+            selection = self.view.sel()
+            if len(selection) != 1 or not selection[0].empty():
+                return "LSP <debug>"  # is_visible will return False
+            pt = selection[0].b
+        for folding_range in sorted(self.folding_ranges, key=lambda r: r['startLine'], reverse=True):
+            region = range_to_region(folding_range_to_range(folding_range), self.view)
+            if region.contains(pt):
+                # Store the relevant folding region, so that we don't need to do the same computation again in
+                # self.is_visible and self.run
+                self.folding_region = region
+                return {
+                    FoldingRangeKind.Comment: "LSP: Fold this comment",
+                    FoldingRangeKind.Imports: "LSP: Fold imports",
+                    FoldingRangeKind.Region: "LSP: Fold this region",
+                    'array': "LSP: Fold this array",  # used by LSP-json
+                    'object': "LSP: Fold this object",  # used by LSP-json
+                }.get(folding_range.get('kind', ''), "LSP: Fold")
+        return "LSP <debug>"  # is_visible will return False
+
+    def run(
+        self,
+        edit: sublime.Edit,
+        manual: bool = True,
+        hidden: bool = False,
+        event: Optional[dict] = None,
+        point: Optional[int] = None
+    ) -> None:
+        if manual:
+            if point is not None:
+                pt = point
+            else:
+                selection = self.view.sel()
+                if len(selection) != 1 or not selection[0].empty():
+                    self.view.run_command('fold_unfold')
+                    return
+                pt = selection[0].b
+            session = self.best_session(self.capability)
+            if session:
+                params = {'textDocument': text_document_identifier(self.view)}  # type: FoldingRangeParams
+                session.send_request_async(
+                    Request.foldingRange(params, self.view),
+                    partial(self._handle_response_manual_async, pt)
+                )
+        elif self.folding_region is not None:
+            self.view.fold(self.folding_region)
```

Comment:
I've been testing a bit and I think we could potentially improve behavior in some cases (but not necessarily in this PR).

What I find a bit inconsistent is when I fold a region and then press the fold shortcut again.

Let's take JSON as an example:

1.

![Screenshot 2023-07-30 at 15 57 32](https://github.com/sublimelsp/LSP/assets/153197/fa838108-053b-471b-9382-80a7481911ed)

2.

![Screenshot 2023-07-30 at 15 57 50](https://github.com/sublimelsp/LSP/assets/153197/89de272e-2da6-4bce-bdcd-1d8852b95960)

At this point pressing the shortcut doesn't do anything. I would expect it to either unfold the folded region or fold  the "outer" region.

(Of course there is no structure to regions so it would be our job to figure out which regions overlap and what would make sense if we'd want to fold "outer" region in this case.)

Reply:
> At this point pressing the shortcut doesn't do anything. I would expect it to either unfold the folded region or fold the "outer" region.

Good point, and I think both alternatives wouldn't be too difficult to implement. When the region is already folded, the search in the for-loop for the relevant region will still match, because the caret is on one side of the region. I just need to add another check in the for-loop if the relevant region is already folded, and then either ignore it (fold "outer" region) or unfold it.

Which of the two option would you prefer / think is better?

Reply:
The ST behavior is to fold the "outer" region if current is already folded so I guess it would make sense to match that.

Another argument for that would be that the action is called "Fold" and not "Toggle fold" so in theory it should never unfold (even when there is non-empty selection).